### PR TITLE
Alerting: Add label and state filters to alert instance components

### LINKED ISSUE #36048: [Feature Request] [NG-Alerting] Filter alert instances based on their state in Alerts view and Alert List panel
**What would you like to be added**:
Ability to filter ng-alert **instances** based on their state, rather than only per alert.
This would be great for both the Alert view as Alert List panel.

**Why is this needed**:
We have a few alerts which thanks to the new alerting in Grafana 8 are multidimensional and create hundrets of alert instances each.
Going through them to find the one that's firing can be quite cumbersome.

image

Version: 8.0.3

### PR DESCRIPTION
**Pull request by JacobsonMT**
**What this PR does / why we need it**:

Adds label and state filtering to all components that use AlertInstancesTable, this includes:

- Alert List panel (this panel was standardized to use AlertInstancesTable)
- Alerting / View rule
- Alerting / Alert rules
- Panel-specific alert tab

Also modifies RadioButton with optional onClick to allow for RadioButtonGroup toggling on second click

**Screenshots / gifs**:

Peek 2021-11-30 18-01

Peek 2021-11-30 18-13

Screenshot from 2021-11-30 18-28-49

**Which issue(s) this PR fixes**:

Fixes #36048
Relates to #35720

### CODE REVIEW

**Comment by JohnnyQQQQ** [comment]
@JacobsonMT thanks for your first contribution to Grafana and welcome to the community 🚀   I did approve the drone build stage and added some folks that will review this frontend change.

**Review by gillesdemey — CHANGES_REQUESTED** [review]
Since this PR is quite large I'm releasing my first batch of comments; some of these are not blockers and can be improved upon in a follow-up PR.

A few blocking issues though, those are not preceded by `nit:` 💡

File: public/app/features/alerting/unified/components/rules/RuleDetailsMatchingInstances.tsx [hunk 1]
```diff
@@ -11,13 +18,126 @@ type Props = {
...
+    (): Alert[] =>
+      filterAlerts(
+        queryString,
+        alertState,
+        sortAlerts(SortOrder.Importance, isAlertingRule(promRule) ? promRule.alerts : [])
+      ),
+    [promRule, alertState, queryString]
+  );
+
+  if (!alerts.length) {
```
[the remark this answers is not in the document]
Comment (JacobsonMT) [reply public/app/features/alerting/unified/components/rules/RuleDetailsMatchingInstances.tsx:40]: Good catch, thanks! That's what I get for not retesting after fixing pre-commit errors :)

File: public/app/features/alerting/unified/components/rules/RuleDetailsMatchingInstances.tsx [hunk 2]
```diff
@@ -11,13 +18,126 @@ type Props = {
 export function RuleDetailsMatchingInstances(props: Props): JSX.Element | null {
   const { promRule } = props;
 
-  if (!isAlertingRule(promRule) || !promRule.alerts?.length) {
```
[the remark this answers is not in the document]
Comment (JacobsonMT) [reply public/app/features/alerting/unified/components/rules/RuleDetailsMatchingInstances.tsx:14]: So, for recording rules we should not display the table/filters at all (return null)?

[the remark this answers is not in the document]
Comment (gillesdemey) [reply public/app/features/alerting/unified/components/rules/RuleDetailsMatchingInstances.tsx:14]: Yep exactly, that is also why the `RuleDetailsMatchingInstances` was typed as `JSX.Element | null` :v:

File: public/app/features/alerting/unified/utils/alertmanager.ts [hunk 3]
```diff
@@ -136,7 +136,7 @@ export function parseMatcher(matcher: string): Matcher {
 }
 
 export function parseMatchers(matcherQueryString: string): Matcher[] {
-  const matcherRegExp = /\b(\w+)(=~|!=|!~|=(?="?\w))"?([^"\n,]*)"?/g;
+  const matcherRegExp = /\b([\w.-]+)(=~|!=|!~|=(?="?\w))"?([^"\n,]*)"?/g;
```
[the remark this answers is not in the document]
Comment (gillesdemey) [reply public/app/features/alerting/unified/utils/alertmanager.ts:139]: Awesome! TIL that the `.` is literal in a character class!

As far as suggestions go I can think of the underscore (`_`) character if we aren't including that one yet, that should do it for now :)

File: packages/grafana-ui/src/components/Forms/RadioButtonGroup/RadioButtonGroup.tsx [hunk 4]
```diff
@@ -23,6 +24,7 @@ export function RadioButtonGroup<T>({
   options,
   value,
   onChange,
+  onClick,
```
[the remark this answers is not in the document]
Comment (JacobsonMT) [reply packages/grafana-ui/src/components/Forms/RadioButtonGroup/RadioButtonGroup.tsx:27]: Implying adding this new "unselect an item by re-clicking it" as the default for RadioButtonGroup? I think it's a good idea, I mostly did it this way as I was a bit wary of modifying the default functionality of something used in so many places of the UI.

I'll give it a shot and see if I notice any strangeness as a result.

**Comment by JacobsonMT** [comment]
Addressed all comments except for "onUnselect" as I'd like to test it a bit more locally

**Review by gillesdemey — APPROVED** [review]
Amazing first-time contribution, LGTM! :shipit:

**Pull request merged**